## HW3

#### Daniel Reynolds 3/31/2021

Useful symbols: ⊗,⊕,$\sqrt{10}$,$F_{base}$,$a^{⊗b}$

In [1]:
%config IPCompleter.greedy=True
from IPython.display import display, Markdown, Latex
#imports
from millerRabin import millerRabin
from fermatFactor import fermatFactor
from binom import binom
from probalisticIsPrime import probalisticIsPrime
from polyAdd import polyAdd
from polyLongDiv import polyLongDiv
from polyMult import polyMult
from powmod import powMod, powmod
from polyPowmod import polyPowmod
from computeInverse import computeInverse
from primitive import primitive, fieldPrimitive
from getCycle import getCycle,getFieldCycle
from sqrtMod import sqrtField
from isASquare import isASquare
from pFactor import pFactor
from allPrimitives import allPrimitives
from euler import euler
from fermatTest import fermatTest
from baseToPoly import baseToPoly
from frobenius import frobeniusPowers,frobeniusSqrt,frobenius
from trace import trace
from quadraticEq import quadraticEq, quadraticNumSolutions
from isASquare import isASquare, legendre, jacobi
from allSquares import allSquares,allSquaresAndRoots
from characteristic import characteristic
from pFactor import pFactor
from bitsToPoly import bitsToPoly
from fieldDiv import fieldDiv
from zeroKnowledgeProof import zeroKnowledgeProof
from divMod import divMod
from chineseRemainderTheorem import crt, printCrtTable, crtJacobi
from gcd import gcd
from generateBigPrime import generateBigPrime

### Problem 1

#### (a) $(\frac{129}{191})$

In [2]:
legendre(129,191)

(129/191) => factoring 129

  (3/191) => * -1
    (191/3) => 
      (2/3) => -1

  (43/191) => * -1
    (191/43) => 
      (19/43) => * -1
        (43/19) => 
          (5/19) => * 1
            (19/5) => 
              (4/5) => factoring 4

                (2/5) => -1

                (2/5) => -1


1

#### 1, which means 129 is a square mod 191

#### (b) $(\frac{795}{1531})$

In [3]:
legendre(795,1531)

(795/1531) => factoring 795

  (3/1531) => * -1
    (1531/3) => 
      (1/3) => 1

  (5/1531) => * 1
    (1531/5) => 
      (1/5) => 1

  (53/1531) => * 1
    (1531/53) => 
      (47/53) => * 1
        (53/47) => 
          (6/47) => factoring 6

            (2/47) => 1

            (3/47) => * -1
              (47/3) => 
                (2/3) => -1


-1

#### -1, which means 795 is not a square mod 1531

### Problem 2

#### (a) $(\frac{41}{79})$

In [4]:
jacobi(41,79)

(41/79) => 2e * 1
  (79/41) => 1c, doing a = 79 % 41
    (38/41) => 1a 38 / 2
      (2/41) => 2d = 1
      (19/41) => 2e * 1
        (41/19) => 1c, doing a = 41 % 19
          (3/19) => 2e * -1
            (19/3) => 1c, doing a = 19 % 3
              (1/3) => 2b = 1


-1

#### -1

#### (b) $(\frac{547}{647})$

In [5]:
jacobi(547,647)

(547/647) => 2e * -1
  (647/547) => 1c, doing a = 647 % 547
    (100/547) => 1a 100 / 2
      (2/547) => 2d = -1
      (50/547) => 1a 50 / 2
        (2/547) => 2d = -1
        (25/547) => 2e * 1
          (547/25) => 1c, doing a = 547 % 25
            (22/25) => 1a 22 / 2
              (2/25) => 2d = 1
              (11/25) => 2e * 1
                (25/11) => 1c, doing a = 25 % 11
                  (3/11) => 2e * -1
                    (11/3) => 1c, doing a = 11 % 3
                      (2/3) => 2d = -1


-1

#### -1

#### (c) $(\frac{419}{461})$

In [6]:
jacobi(419,461)

(419/461) => 2e * 1
  (461/419) => 1c, doing a = 461 % 419
    (42/419) => 1a 42 / 2
      (2/419) => 2d = -1
      (21/419) => 2e * 1
        (419/21) => 1c, doing a = 419 % 21
          (20/21) => 1a 20 / 2
            (2/21) => 2d = -1
            (10/21) => 1a 10 / 2
              (2/21) => 2d = -1
              (5/21) => 2e * 1
                (21/5) => 1c, doing a = 21 % 5
                  (1/5) => 2b = 1


-1

#### -1

### Problem 3

#### (a)

Methodology:
<br>
If we know b will be 0, pick a random z and set $y=z^{2}$ mod n.
<br>
If we know b will be 1, pick a random z and set $y=\frac{z^{2}}{x}$ mod n.

In [7]:
# My work:
n = 58014043
x = 18059241
def test(y,b,z,i):
    print(f"Using y={y} and z={z} for a b={b}")
    if zeroKnowledgeProof(y,b,z,n,x):
        print(f"Bob accepts @i={i}\n")
    

z1 = 12345 
y1 = powmod(z1,2,n) 
test(y1,0,z1,1)

z2 = 12345
y2 = divMod(powmod(z2,2,n),x,n)
test(y2,1,z2,2)

z3 = 6789
y3 = divMod(powmod(z3,2,n),x,n)
test(y3,1,z3,2)

z4 = 6789
y4 = powmod(z4,2,n) 
test(y4,0,z4,4)

Using y=36370939 and z=12345 for a b=0
Bob accepts @i=1

Using y=21431605 and z=12345 for a b=1
Bob accepts @i=2

Using y=16485723 and z=6789 for a b=1
Bob accepts @i=2

Using y=46090521 and z=6789 for a b=0
Bob accepts @i=4



|$i$|$y_{i}$|$b_{i}$|$z_{i}$|Bob Checks|Accept|
|:--|:--    |:--    |:--    |:--       |:--   |
|1|36370939|0|12345|$z^{2}\equiv y$ mod $n$|Yes|
|2|21431605|1|12345|$z^{2}\equiv xy$ mod $n$|Yes|
|3|16485723|1|6789|$z^{2}\equiv xy$ mod $n$|Yes|
|4|46090521|0|6789|$z^{2}\equiv y$ mod $n$|Yes|

#### (a)

Brute force to find $s$

In [8]:
n = 58014043
x = 8952064
sFound = False
s = 1
while not(sFound):
    s = s+1
    if powmod(s,2,n) == x:
        print(f"Found s={s}")
        sFound = True
s

Found s=2992


2992

$s=2992$

In [9]:
# we know s=2992, lets do the game

u1 = 24681
y1 = powmod(u1,2,n)
# we recieve a 1 from bob
z1 = (u1 * s) % n
test(y1,1,z1,1)

u2 = 67891
y2 = powmod(u2,2,n)
# we recieve a 1 from bob
z2 = (u2 * s) % n
test(y2,1,z2,2)

u3 = 67891
y3 = powmod(u3,2,n)
# we recieve a 0 from bob
z3 = u3
test(y3,0,z3,3)

u4 = 3927
y4 = powmod(u4,2,n)
# we recieve a 0 from bob
z4 = u4
test(y4,0,z4,4)

Using y=29011331 and z=15831509 for a b=1
Bob accepts @i=1

Using y=26078484 and z=29087743 for a b=1
Bob accepts @i=2

Using y=26078484 and z=67891 for a b=0
Bob accepts @i=3

Using y=15421329 and z=3927 for a b=0
Bob accepts @i=4



|$i$|$u_{i}$|$y_{i}$|$b_{i}$|$z_{i}$|Bob Checks|Accept|
|:--|:--    |:--    |:--    |:--    |:--       |:--   |
|1|24681|29011331|1|15831509|$z^{2}\equiv xy$ mod $n$|Yes|
|2|67891|26078484|1|29087743|$z^{2}\equiv xy$ mod $n$|Yes|
|3|67891|26078484|0|67891|$z^{2}\equiv y$ mod $n$|Yes|
|4|3927|15421329|0|3927|$z^{2}\equiv y$ mod $n$|Yes|

### Problem 4

In [10]:
p = 3
q = 7

#### (a)

In [11]:
printCrtTable(crt(3,7))

|p/q|$0$|$1$|$2$|$3$|$4$|$5$|$6$|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|$0$|$0$|$15$|$9$|$3$|$18$|$12$|$6$|
|$1$|$7$|$1$|$16$|$10$|$4$|$19$|$13$|
|$2$|$14$|$8$|$2$|$17$|$11$|$5$|$20$|

#### (b)

In [12]:
printCrtTable(crtJacobi(3,7))

|p/q|$0$|$1$|$2$|$3$|$4$|$5$|$6$|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|$0$|$0$|$0$|$0$|$0$|$0$|$0$|$0$|
|$1$|$0$|$1$|$1$|$-1$|$1$|$-1$|$-1$|
|$2$|$0$|$-1$|$-1$|$1$|$-1$|$1$|$1$|

#### (c)

In [25]:
allSquaresAndRoots(21)

Finding all squares and their roots in F21


{0: [0],
 1: [1, 8, 13, 20],
 4: [2, 5, 16, 19],
 9: [3, 18],
 16: [4, 10, 11, 17],
 15: [6, 15],
 7: [7, 14],
 18: [9, 12]}

|p/q|$0$|$1$|$2$|$3$|$4$|$5$|$6$|
|:--|:--|:--|:--|:--|:--|:--|:--|
|$0$|✓|✓|✓|✘|✓|✘|✘|
|$1$|✓|✓|✓|✘|✓|✘|✘|
|$2$|✘|✘|✘|✘|✘|✘|✘|

#### (d)

#### There are 8 elements that are squares mod 21, if you include 0.